presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,369 @@
+                    .forEach(allocationCounts::add);
+
+            int currentMin = allocationBytes.keySet().stream()
+                    .map(allocationCounts::count)
+                    .min(Integer::compare)
+                    .get();
+            int currentMax = allocationBytes.keySet().stream()
+                    .map(allocationCounts::count)
+                    .max(Integer::compare)
+                    .get();
+
+            int numBuckets = distributionAssignments.size();
+            int targetMin = (int) Math.floor((numBuckets * 1.0) / clusterState.getActiveNodes().size());
+            int targetMax = (int) Math.ceil((numBuckets * 1.0) / clusterState.getActiveNodes().size());
+
+            log.info("Distribution %s: Current bucket skew: min %s, max %s. Target bucket skew: min %s, max %s", distribution.getId(), currentMin, currentMax, targetMin, targetMax);
+
+            for (String source : ImmutableSet.copyOf(allocationCounts)) {
+                List<BucketAssignment> existingAssignments = distributionAssignments.stream()
+                        .filter(assignment -> assignment.getNodeIdentifier().equals(source))
+                        .collect(toList());
+
+                for (BucketAssignment existingAssignment : existingAssignments) {
+                    if (activeNodes.contains(source) && allocationCounts.count(source) <= targetMin) {
+                        break;
+                    }
+
+                    // identify nodes with bucket counts lower than the computed target, and greedily select from this set based on projected disk utilization.
+                    // greediness means that this may produce decidedly non-optimal results if one looks at the global distribution of buckets->nodes.
+                    // also, this assumes that nodes in a cluster have identical storage capacity
+                    String target = activeNodes.stream()
+                            .filter(candidate -> !candidate.equals(source) && allocationCounts.count(candidate) < targetMax)
+                            .sorted((a, b) -> Integer.compare(allocationCounts.count(a), allocationCounts.count(b)))
+                            .min((a, b) -> Double.compare(allocationBytes.get(a), allocationBytes.get(b)))
+                            .orElseThrow(() -> new RuntimeException("unable to find target for rebalancing"));
+
+                    long bucketSize = clusterState.getDistributionBucketSize().get(distribution);
+
+                    // don't move bucket if it doesn't reduce imbalance
+                    if (activeNodes.contains(source) && (allocationCounts.count(source) == targetMax && allocationCounts.count(target) == targetMin)) {
+                        break;
+                    }
+
+                    allocationCounts.remove(source);
+                    allocationCounts.add(target);
+                    allocationBytes.compute(source, (k, v) -> v - bucketSize);
```

Comment:
This invocation is not a good candidate for merge - there is no 'default' value.